### PR TITLE
New rule: Visible focus

### PR DESCRIPTION
**Pull request by Jym77**
New rule for 2.4.7: Visible Focus.

The rule uses the "distinguishing styles" definition from https://github.com/act-rules/act-rules.github.io/blob/inline_link_in_paragraph_is_distinguishable/pages/glossary/distinguishing-styles.md and will need to augment it a bit.

The rule currently has an outstanding issue of not handling correctly the default focus indicator of input fields (blinking cursor) and I am not sure how to solve it (ideas welcome!)

Closes issue(s):
N/A

Need for Final Call:
This will require a 2 weeks Final Call (new rule).

---

### **When creating PR:**

- [X] Make sure you're requesting to **pull a branch** (right side) to the `develop` branch (left side).

### **After creating PR:**

- [X] Add yourself (and co-authors) as "Assignees" for PR.
- [X] Add label to indicate if it's a `Rule`, `Definition` or `Chore`.
- [ ] [Link the PR to any issue it solves](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue). This will be done automatically by [referencing the issue](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) at the top of this comment in the indicated place.
- [ ] Optionally request feedback from anyone in particular by assigning them as "Reviewers".

### **When merging a PR:**

- [ ] Close any issue that the PR resolves. This will happen automatically upon merging if the PR was correctly linked to the issue, e.g. by referencing the issue at the top of this comment.

## How to Review And Approve

- Go to the “Files changed” tab
- Here you will have the option to leave comments on different lines.
- Once the review is completed, find the “Review changes” button in the top right, select “Approve” (if you are really confident in the rule) or "Request changes" and click “Submit review”.
- Make sure to also review the proposed Final Call period. In case of disagreement, the longer period wins.

### CODE REVIEW

File: _rules/focusable-element-has-visible-focus-f4e323.md
Comment (WilcoFiers) [inline _rules/focusable-element-has-visible-focus-f4e323.md:27]: I think you're going to have a very hard time to come up with a complete list of all the different ways something can be focusable. Perhaps focus indicator can be defined as all pixels that change when the element is focused from when the element is not focused.

You could add a second expectation that says each focus indicator must be unique, to replicate this "for another focusable element" thing, but I'd much prefer to have that as an assumption. I don't think I've ever seen that happen, so that seems like a hypothetical we shouldn't spend our time on.

File: _rules/focusable-element-has-visible-focus-f4e323.md
Comment (Jym77) [reply _rules/focusable-element-has-visible-focus-f4e323.md:23]: > The correct wording here should probably have been _"where the keyboard focus has a visible indicator"_.

I do not understand what you mean here. If the keyboard focus indicator is visible, then the SC is satisfied and there is nothing to test 🤔 


That said, I do agree that restricting the rule to sequential focus navigation makes a lot of sense 👍

File: _rules/focusable-element-has-visible-focus-f4e323.md
Comment (Jym77) [reply _rules/focusable-element-has-visible-focus-f4e323.md:27]: I disagree with the need for unicity. Check Passed Example 6 where each element has 2 focus indicator (one on each side) and these focus indicators are shared between two elements. The only unicity condition that applies to this example is that the set of stuff that change is unique.

File: pages/glossary/distinguishing-styles.md
Comment (Jym77) [reply pages/glossary/distinguishing-styles.md:11]: Outline is included through https://github.com/act-rules/act-rules.github.io/blob/inline_link_in_paragraph_is_distinguishable/pages/glossary/distinguishing-borders.md 😁 

I admit I haven't followed all discussions in that other rule, but I kinda assume that whatever "visible focus" that one ends up with can be reused here (as in "if it's good there, it should be good here").
Granted, 2.4.7 is very lax in how focus should be shown ("visible", essentially), but 1.4.1 (for which the other rule and def are worked) is also super lax ("color is not the only way").

File: _rules/focusable-element-has-visible-focus-f4e323.md
Comment (kasperisager) [inline _rules/focusable-element-has-visible-focus-f4e323.md:27]:
```suggestion
For each target element, there exists a set of [focus indicators][focus indicator] which are not all also [focus indicators][focus indicator] for another [focusable][] element.
```

It seemed a little detached 🤔

Comment (Jym77) [reply _rules/focusable-element-has-visible-focus-f4e323.md:27]: The precision is needed essentially for cases like Passed Example 7 where there are several focusable elements in a row and focus is indicated by lighting a dot on both side of each focusable. Then, each focus indicator is a focus indicator for 2 different elements. So for elements in the middle, all of their focus indicators are also focus indicator for another element, however this is not the same other element for all focus indicator, so this is good.

File: _rules/sequentially-focusable-element-has-visible-focus-through-styling-f4e323.md
Comment (Jym77) [inline _rules/sequentially-focusable-element-has-visible-focus-through-styling-f4e323.md:3]: Note to reviewers:
I initially wrote this rule that is too strict for WCAG (2.4.7 is notably weak in its requirements). After discussing with @WilcoFiers, we decided to go for the "minimal version" (which is the other rule).

I was reluctant about throwing away all the work here 😉 So I decided to try and keep it as a "best practice" rule (it doesn't really check for 2.4.11 either). Let me know what you think about it and whether we should keep it or not…
(if we keep it, we'll probably want to move it to another PR for separation of concerns, of course, I'll do that if initial feedback on it is good 😀)

File: _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md
Comment (carlosapaduarte) [inline _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md:100]:
```suggestion
<script src="../test-assets/focus-visible/script.js"></script>
```

File: _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md
Comment (carlosapaduarte) [inline _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md:96]:
```suggestion
The first [focusable][] element, is part of [sequential focus navigation][]. The set of pixels that change color when it is [focused][] (the border of the `span` element) is unique (it does not change when any other element is [focused][]). The second element part of [sequential focus navigation][] also as a unique set of color changing pixels when [focused][].
```

File: _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md
Comment (carlosapaduarte) [inline _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md:121]:
```suggestion
<script src="../test-assets/focus-visible/script.js"></script>
```

File: _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md
Comment (carlosapaduarte) [inline _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md:31]:
```suggestion
For each target element and each other [focusable][] element in the document, the set of pixels whose color is different when the test target is [focused][] and when it is not, and the set of pixels whose color is different when the other element is [focused][] and when it is not, are different.
```

File: _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md
Comment (carlosapaduarte) [inline _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md:162]:
```suggestion
```html
<link rel="stylesheet" href="../test-assets/focus-visible/styles.css" />

<a class="no-focus-default" href="https://act-rules.github.io/">ACT rules</a>
<a class="no-focus-default" href="https://www.w3.org/TR/WCAG21/">WCAG</a>
```
```

File: _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md
Comment (carlosapaduarte) [inline _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md:157]:
```suggestion
None of these [focusable][] elements have any pixel changing color when they are [focused][] because the default styling has been overwritten by a style that removes the outline.
```

File: _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md
Comment (carlosapaduarte) [inline _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md:171]:
```suggestion
```html
<link rel="stylesheet" href="../test-assets/focus-visible/styles.css" />

<a class="no-focus-default" href="https://act-rules.github.io/">ACT rules</a>
<button tabindex="-1">Dummy button</button>
```
```

File: _rules/sequentially-focusable-element-has-visible-focus-through-styling-f4e323.md
Comment (carlosapaduarte) [reply _rules/sequentially-focusable-element-has-visible-focus-through-styling-f4e323.md:3]: I've just browsed this rule, not really reviewed it (because, as you say, we should move it to a different PR if we decide to keep it).

I feel your pain, because of all the effort you must have put into this, but I'm not really sure it will add much value to the other visible focus rule.

File: _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md
Comment (Jym77) [reply _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md]: I tried to be explain this.
It was surprisingly difficult to say "somewhere on the page", and I am still not sure what I picked is correct (notably for infinite scrolling pages like maps)…

File: _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md
Comment (Jym77) [inline _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md:8]: 🤔 
Actually, 2.4.7 is only level AA in WCAG 2.1 but will move to level A in WCAG 2.2.

Not sure how to handle that…

File: _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md
Comment (WilcoFiers) [inline _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md:165]: Looks like you broke prettier

File: _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md
Comment (Jym77) [reply _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md:31]: I feel that the understanding document is strongly pushing toward unicity with sentences like:
> It must be possible for a person to know which element among multiple elements has the keyboard focus.

if multiple elements have the same focus indicator, it is not possible to know which among them has focus.

> - This Success Criterion helps anyone who relies on the keyboard to operate the page, by letting them visually determine the component on which keyboard operations will interact at any point in time.

uses singular "**the** component" none "some components, one of which has focus".

> - People with attention limitations, short term memory limitations, or limitations in executive processes benefit by being able to discover where the focus is located.

With short term memory, if you can't remember where on the page you are, you'll need focus indicator to be unique in order to be somewhat helpful.

---
I'd rather keep this expectation and add an assumption that unicity is required by WCAG.

File: _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md
Comment (Jym77) [reply _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md]: While not technically normative, the Understanding document still says:
> The Working Group considers the content to be authoritative.

But sure, we can make an assumption that the Understanding document is correct.

File: _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md
Comment (Jym77) [inline _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md:22]: Note that this is currently not handled by the build. We'll need to grab WCAG 2.2 data, which is currently not done…

Comment (WilcoFiers) [reply _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md:22]: Yeah, I'd leave this off until we can do the full migration to WCAG 2.2.

File: _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md
Comment (WilcoFiers) [reply _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md:41]: > It must be possible for a person to know which element among multiple elements has the keyboard focus.

That's all well and good, but if it's not in the SC text, then it's not a normative requirement. I'm also not convinced this is something that actually happens. I've definitely never seen an issue like this. The only scenario I can come up with for this is the one above, where I'd believe this creates a false positive.

If you'd like to discuss this on an upcoming call, feel free to put the agenda item label on it, but I don't see a reason to change my vote on this.

**Review by WilcoFiers — DISMISSED** [review]

File: _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md
Comment (WilcoFiers) [inline _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md:58]: I don't think this "at least one pixel" phrase needs to be in these descriptions. We don't want to repeat the expectation. I think the following would be sufficient:

```suggestion
The first [focusable][] element is part of [sequential focus navigation][]. It gets an outline focused because of the User Agent's default styling. The second [focusable][] element is not applicable because it has been removed from [sequential focus navigation][] due to the negative value for the `tabindex` attribute. Its presence is nonetheless enough to make the first one applicable.
```

The same comment applies to other exampled.

**Review by WilcoFiers — CHANGES_REQUESTED** [review]

**Review by WilcoFiers — APPROVED** [review]

File: _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md
Comment (carlosapaduarte) [inline _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md:23]: To make sure this applies to HTML documents

```suggestion
The rule applies to any element which is part of [sequential focus navigation][] in an [HTML document](https://html.spec.whatwg.org/#document) with at least two [focusable][] elements.
```

File: _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md
Comment (carlosapaduarte) [inline _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md:31]:
```suggestion
This rule assumes that documents with only one focusable element automatically satisfy [Success Criterion 2.4.7 Focus Visible][sc247], as stated in [Understanding Success Criterion 2.4.7: Focus Visible][usc247]: "if there is only one keyboard actionable control on the screen, the success criterion would be met".
```

File: _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md
Comment (carlosapaduarte) [inline _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md:41]: Not sure what you're meaning by a "too weak" focus indicator... is it perhaps contrast related?

```suggestion
WCAG does not have any requirement of how big or small focus indicator should be, or how far or near from the [focusable][] element it should be. Thus it is possible to pass this rule and [Success Criterion 2.4.7 Focus Visible][sc247] with barely perceptible changes at the other end of the page. That would however still be an accessibility issue. WCAG 2.2 includes [Success Criterion 2.4.11 Focus Appearance (Minimum)][sc2411] and [Success Criterion 2.4.12 Focus Appearance (Enhanced)][sc2412] specifying how big the focus indicator should be. All Examples in this rule avoid focus indications that are too small, too far away, have low contrast, …
```

File: _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md
Comment (carlosapaduarte) [inline _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md:58]:
```suggestion
The first [focusable][] element is part of [sequential focus navigation][]. It has an outline when it is [focused][] (due to default User Agent's styling). The second [focusable][] element is not applicable because it has been removed from [sequential focus navigation][] due to the negative value for the `tabindex` attribute. Its presence is nonetheless enough to make the rule applicable to the first one.
```

File: _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md
Comment (carlosapaduarte) [inline _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md:164]:
```suggestion
None of the [focusable][] elements in this document are part of [sequential focus navigation][].
```

File: _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md
Comment (carlosapaduarte) [inline _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md:135]: I would place the button in the sequential focus order (i.e. remove the negative tabindex) to make it clear that even if one of the elements pass the rule, it fails because all need to pass it

**Review by carlosapaduarte — CHANGES_REQUESTED** [review]
Some minor requests and typos to fix

File: _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md
Comment (Jym77) [reply _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md:135]: I'm afraid that mixing Passed and Failed targets in the same test case will be messing up the implementation reports.
Currently, we expect one "failed" result for this test case. If we make the other one passing, we'd expect one "fail" and one "passed" result, and thus we also need to figure out which outcome is for which target in order to decide if the implementation is correct, and that makes automated implementations mapping harder, I guess…

File: _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md
Comment (carlosapaduarte) [inline _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md:74]:
```suggestion
The first [focusable][] element is part of [sequential focus navigation][]. Its parent `span` puts a border around it when it is [focused][] (the border of the `span` element).
```

File: _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md
Comment (carlosapaduarte) [reply _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md:135]: That's fair

**Review by carlosapaduarte — CHANGES_REQUESTED** [review]
Still found one typo

File: _rules/object-has-acessible-name-8fc3b6.md
Comment (Jym77) [reply _rules/object-has-acessible-name-8fc3b6.md]: Because #1488 wasn't merged in yet…

File: _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md
Comment (kasperisager) [inline _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md:31]: The keyword here seems to be "keyboard actionable". It therefore seems like there's a disconnect between this assumption and the applicability of the rule as the latter requires at least two "focusable" elements in the document, not at least two "keyboard actionable" elements.

**Review by kasperisager — DISMISSED** [review]

File: _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md
Comment (carlosapaduarte) [reply _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md:31]: And given that @Jym77 has already created the definition for [keyboard actionable](https://github.com/act-rules/act-rules.github.io/blob/bf57d85c3f0514721dc70868506b3519a2e65bda/pages/glossary/keyboard-actionable-element.md) that should be an easy fix 🙂

**Review by carlosapaduarte — COMMENTED** [review]
Will approve after the changes suggested by @kasperisager are put in place

File: _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md
Comment (Jym77) [reply _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md:31]: Interesting discussion…
I am not convinced that our "keyboard actionable" make sense here. If an element can be tabbed but not clicked, it still require focus indication. Typically, [HTML specs](https://html.spec.whatwg.org/multipage/interaction.html#data-model:sequentially-focusable) have an example where a UA could make bullets (the bullet symbol) in a list could be sequentially focusable as a reading guide, in which case they definitely need an indicator…

OTOH, WCAG (the SC) is only about "a keyboard operable UI component"… Not sure whether these bullets would be considered as a UI component by WCAG… But "operable" could be understand as "tabbable", if the only purpose is to be tabbable (reading guide, …)

Suggestion: put the applicability as "if there are at least 2 sequentially focusable elements" (rather than "at least 2 focusable)"; and add an assumption "this rule assumes that sequentially focusable elements are UI components", which sounds safe.
+ update all examples since the dummy button with `tabindex=-1` does not work any more 😕 

This would also avoid the need for "keyboard actionable" which is annoying to test automatically. So the Applicability could be automated, and the rule also in many cases (e.g., detecting an indicator through styling, …)

@carlosapaduarte @kasperisager WDYT?

Comment (kasperisager) [reply _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md:31]: Something along the lines of "if the sequential focus navigation order contains more than 1 element" works fine by me 👍

Comment (carlosapaduarte) [reply _rules/sequentially-focusable-element-has-visible-focus-oj04fd.md:31]: Agreed

**Review by carlosapaduarte — APPROVED** [review]

**Review by kasperisager — APPROVED** [review]

**Comment by Jym77** [comment]
Final Call ends on November 11.

**Comment by Jym77** [comment]
Final Call has ended. Merging.

**Pull request merged**
